### PR TITLE
Tonight, I drink from the keg of glory [Goodbye, Keep.libraryId]

### PR DESCRIPTION
**Pull request by ryanpbrewster**
:triumph: 

@andrewconner @camhashemi @leogrim

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
I got so lazy/frustrated towards the end of this PR. `Keep.lowestLibraryId` is my secret shame.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala [hunk 1]
```diff
@@ -789,53 +779,33 @@ class KeepCommanderImpl @Inject() (
...
-        ktlCommander.internKeepInLibrary(movedKeep, toLibrary, Some(userId))
-
-        Right(movedKeep)
-      case Some(existingKeep) =>
-        if (toLibrary.id.get == k.libraryId.get) {
-          Left(LibraryError.AlreadyExistsInDest)
-        } else {
+        ktlCommander.removeKeepFromAllLibraries(k.id.get)
+        ktlCommander.internKeepInLibrary(k, toLibrary, addedBy = Some(userId))
+        println(s"just moved keep ${k.id.get} into lib ${toLibrary.id.get}")
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala]: :1234:

**Comment by camhashemi** [comment]
:+1: lgtm minus some `println`s. EPIC PR

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala]: :smirk_cat: sneaky bastard

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala [hunk 2]
```diff
@@ -122,8 +109,7 @@ object Keep extends PublicIdGenerator[Keep] {
     (__ \ 'keptAt).format[DateTime] and
     (__ \ 'lastActivityAt).format[DateTime] and
     (__ \ 'messageSeq).formatNullable[SequenceNumber[Message]] and
-    (__ \ 'connections).format[KeepConnections] and
-    (__ \ 'libraryId).formatNullable[Id[Library]]
+    (__ \ 'connections).format[KeepConnections]
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala:112]: Any consequence on APIs or caching here?

File: kifi-backend/modules/common/conf/evolutions/shoebox/432.sql [hunk 3]
```diff
@@ -2,9 +2,14 @@
 
 # --- !Ups
 
+ALTER TABLE bookmark DROP CONSTRAINT bookmark_f_library_id;
+ALTER TABLE bookmark DROP CONSTRAINT lib_primary_uri;
+DROP INDEX bookmark_i_library_id_kept_at;
```
Comment (leogrim) [inline kifi-backend/modules/common/conf/evolutions/shoebox/432.sql:7]: I'm not super comfortable dropping these constraints until we have a new interning logic in place at keep creation

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala [hunk 4]
```diff
@@ -459,11 +458,11 @@ class KeepCommanderImpl @Inject() (
...
-    keeps.groupBy(_.libraryId).toList.collect {
-      case (Some(libId), _) =>
-        val library = libraryRepo.get(libId)
-        libraryRepo.save(library.copy(keepCount = keepRepo.getCountByLibrary(libId)))
-        libraryAnalytics.unkeptPages(userId, keeps, library, context)
+    val libIds: Set[Id[Library]] = keeps.flatMap(_.connections.libraries).toSet
+    libIds.foreach { libId =>
+      val library = libraryRepo.get(libId)
+      libraryRepo.save(library.copy(keepCount = ktlRepo.getCountByLibraryId(libId)))
+      libraryAnalytics.unkeptPages(userId, keeps, library, context)
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala:465]: So unkeeping for now means deleting correct? Or unkeep from all?

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 5]
```diff
@@ -545,89 +496,11 @@ class KeepRepoImpl @Inject() (
...
-      LIMIT $limit
-    """.as[(Id[Keep], DateTime)].list
-
-    val shouldFilterByUser = filterOpt.contains(OwnKeeps)
-    val keepIds = keepsAndFirstAddedAt.map { case (keepId, _) => keepId }
-    val keepsById = getByIds(keepIds.toSet)
-    keepsAndFirstAddedAt.map { case (keepId, firstAddedAt) => keepsById(keepId) -> firstAddedAt }
-      .filter { case (keep, _) => !shouldFilterByUser || keep.userId.contains(userId) }
-  }
-
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:630]: :crying_cat_face:

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 6]
```diff
@@ -693,22 +566,19 @@ class KeepRepoImpl @Inject() (
 
   def getMaxKeepSeqNumForLibraries(libIds: Set[Id[Library]])(implicit session: RSession): Map[Id[Library], SequenceNumber[Keep]] = {
     import com.keepit.common.db.slick.StaticQueryFixed.interpolation
-    if (libIds.isEmpty) {
-      Map()
-    } else {
+    if (libIds.isEmpty) Map.empty
+    else {
       val idset = libIds.map { _.id }.mkString("(", ",", ")")
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:571]: Weird never seen `mkString` with 3 parameters

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala [hunk 7]
```diff
@@ -56,11 +59,19 @@ trait KeepToLibraryRepo extends Repo[KeepToLibrary] {
...
 
-  override def deleteCache(ktl: KeepToLibrary)(implicit session: RSession) {}
-  override def invalidateCache(ktl: KeepToLibrary)(implicit session: RSession) {}
+  override def deleteCache(ktl: KeepToLibrary)(implicit session: RSession): Unit = {
+    countByLibraryCache.remove(CountByLibraryKey(ktl.libraryId))
+    libraryMetadataCache.remove(LibraryMetadataKey(ktl.libraryId))
+  }
+  override def invalidateCache(ktl: KeepToLibrary)(implicit session: RSession): Unit = {
+    // TODO(ryan): is it necessary to actually delete the cache here?
+    deleteCache(ktl)
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:73]: Do you mean in general or the contest of that specific repo?

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala [hunk 8]
```diff
@@ -117,6 +128,9 @@ class KeepToLibraryRepoImpl @Inject() (
     val res = q.as[(Long, Int)].list
     res.map { case (orgId, count) => Id[Organization](orgId) -> count }
   }
+  def countNonImportedKeepsInOrg(orgId: Id[Organization])(implicit session: RSession): Int = {
+    activeRows.filter(_.organizationId === orgId).groupBy(_.keepId).length.run
+  }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:133]: How are these specifically non imported?

**Comment by leogrim** [comment]
The only part I'm unclear on is whether this preserves current keep interning behavior for existing APIs?

**Comment by leogrim** [comment]
fuck yeah

**Comment by eishay** [comment]
lol

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala (hunk 2 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala:112]: Some internal APIs will break until both services are finished deploying. The caches are "backwards compatible" in that they can read the old formats, but mid-deploy the old services may run into trouble. I'll bump the version number so it's a little quieter.

File: kifi-backend/modules/common/conf/evolutions/shoebox/432.sql (hunk 3 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/conf/evolutions/shoebox/432.sql:7]: I can keep them around, but any new keeps are going to have a `NULL` anyways.

There is a similar constraint on `keep_to_library (library_id, is_primary, uri_id)` in place.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala (hunk 4 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepCommander.scala:465]: Existing clients universally intend to delete a keep when they call "unkeep" APIs. Going forward I'd like to repurpose that word to mean "remove from library", but it's not particularly important to me. 

The `/client/keeps/:keepId/recipients` endpoint can remove a keep from a library.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 6 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:571]: Very useful

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala (hunk 7 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:73]: In this context specifically. I _think_ we never change the `keepId` or `libraryId` for a model, so "invalidating" will only happen when we do things like change the URI or bump the `lastActivityAt` (in which case those caches are not affected). Dangerous assumption to bake into the repo, though, I think.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala (hunk 8 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:133]: I have no idea. I think this was copy-pasted from `KeepRepo`.

**Comment by ryanpbrewster** [comment]
If we assume that all keeps are in `{0, 1}` library then this preserves existing behavior, as we are now interning by `(Id[NormalizedURI], Set[Id[Library]])` instead of `(Id[NormalizedURI], Option[Id[Library]])`.

**Comment by leogrim** [comment]
yolo

**Pull request merged**